File: src/main/java/frc/team3324/robot/wrappers/PIDCommand.java
package frc.team3324.robot.wrappers;

import edu.wpi.first.wpilibj.Notifier;
import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.command.Subsystem;

import java.util.function.DoubleConsumer;
import java.util.function.DoubleSupplier;

public class PIDCommand extends Command {

    protected double kP;
    protected double kI;
    protected double kD;
    protected double goal;
    protected double dt;

    protected DoubleSupplier measurement;
    protected DoubleConsumer useOutput;

    protected double lastPosition = 0;
    protected double integral = 0;

    protected Notifier notifier = new Notifier(this ::executePID);

    public PIDCommand(double kP, double kI, double kD, double goal, double dt, Subsystem subsystem, DoubleSupplier measurement,
                      DoubleConsumer useOutput) {
        this.kP = kP;
        this.kI = kI;
        this.kD = kD;

        this.goal = goal;
        this.dt = dt;

        this.requires(subsystem);
        this.measurement = measurement;
        this.useOutput = useOutput;
    }

    @Override
    protected void initialize() {
        integral = 0;
        notifier.startPeriodic(dt);
    }

    protected void executePID() {
        double position = measurement.getAsDouble();
        double error = goal - position;
        double deriv = position - lastPosition;
        integral += error;
        lastPosition = position;

        double output = (error * kP) + (integral * kI) - (deriv * kD);
        useOutput.accept(output);
    }

    @Override
    protected void end() {
        stopNotifier();
    }

    @Override
    protected void interrupted() {
        end();
    }

    public void stopNotifier() {
        notifier.stop();
        notifier.stop();
    }

    @Override
    protected boolean isFinished() {
        return measurement.getAsDouble() == goal;
    }
}

File: src/main/kotlin/frc/team3324/robot/util/PIDCommand.kt
package frc.team3324.robot.util

import edu.wpi.first.wpilibj.Notifier
import edu.wpi.first.wpilibj.command.Command
import edu.wpi.first.wpilibj.command.Subsystem
import java.util.function.DoubleConsumer
import java.util.function.DoubleSupplier

class PIDCommand(val kP: Double, val kI: Double, val kD: Double, val goal: Double, val dt: Double, val subsystem: Subsystem, val measurement: () -> Double, val useOutput: (Double) -> Unit): Command() {
    protected var integral = 0.0
    protected var lastPosition = 0.0
    private val notifier = Notifier(this ::executePID)

    init {
        this.requires(subsystem)
    }

    override fun initialize() {
       integral = 0.0
       notifier.startPeriodic(dt)
    }

    fun executePID() {
        val position = measurement()
        val error = goal - position
        val deriv = position - lastPosition

        integral += error
        lastPosition = position

        val output = (error * kP) + (integral * kI) - (deriv * kD)
        useOutput(output)
    }

    override fun end() {
        stopNotifier()
    }

    override fun interrupted() {
        stopNotifier()
        end()
    }

    fun stopNotifier() {
        notifier.stop()
        notifier.stop()
    }

    override fun isFinished(): Boolean {
        return measurement() == goal
    }

}